### ISSUE TITLE
%hierarchy directive seems not to work

### ISSUE THREAD

**Issue by apblack** [ThierryGoubier/SmaCC#48]
[Grammar-with-hierarchy.txt](https://github.com/ThierryGoubier/SmaCC/files/1450829/Grammar-with-hierarchy.txt)
I've been having trouble with the %hierarchy directive.  Starting with a working grammar, I've been adding %hierarchy directives to create a class hierarchy, before writing a visitor.

I was assuming that any class not listed in a `%hierarchy` directive would be a subclass of the root class, but in fact no code was being generated for some classes at all.  So I added an explicit %hierarchy directive for root.

SmaCC errors with `#includeClass:` sent to `nil`.   The problem appears to arise when calculating the `commonSupertypeOf:` in `SmaCCRHS`.  Here are my hierarchy definitions:

```
%hierarchy Declaration (MethodDeclaration Import Dialect VarDeclaration DefDeclaration ClassOrTraitDeclaration Parameters) ;
%hierarchy Parameters (SingleMethodParameter MethodParameters BlockParameter TypeParameters) ;
%hierarchy Expression (Request Block Factor Term Identifier Object Numeral String Sequence Oneself Boolean) ;
%hierarchy Numeral (DecimalNumeral BaseExponentNumeral BasedRadixNumeral) ;
%hierarchy String (StringLiteral StringConstructor) ;
%hierarchy Statement (Expression Pragma Declaration Assignment Return Import Dialect Empty) ;
%hierarchy Oneself (Self Outer) ;
%hierarchy Request (RequestWithArguments SelfRequest OuterRequest TargetedRequest AssignmentRequest UnaryRequest BinaryRequest) ;
%hierarchy Object (Module) ;
%hierarchy Root (Declaration Parameters Expression Numeral String Statement Oneself Request Object) ;
```
Poking around in the model, the supertype of `GraceRequestNode`, which ought to be `GraceRootNode`, is an instance of `Object`.  

I tried re-ordering the `%hierarchy` statements, wondering if they needed to be bottom up.  (The `%hierarchy` for `Root` used to be at the top).  The superclass of `GraceRequestNode` is still an instance of `Object`(but the superclass of `GraceStatementNode` is — correctly —`GraceRootNode`).

**Comment by apblack** [comment]
It turns out that the cause of this bug is having a previously-compiled set of node classes present in the image.  I discovered this when I reverted to an older, previously-working grammar, only to get the same error.

What worked for me was to manually remove all of the previously-generated code from the system.  I added these two methods to the class-side of my root node class (`GraceRootNode`):

```
removeAllGenerated
	"self removeAllGenerated"
	
	GraceRootNode allSubclassesDo: [  :c | c removeFromSystem ].
	self removeGeneratedMethodsFrom: GraceRootNode.
	self removeGeneratedMethodsFrom: (Smalltalk globals at: #GraceParser).
	self removeGeneratedMethodsFrom: (Smalltalk globals at: #GraceParser) class.
	self removeGeneratedMethodsFrom: (Smalltalk globals at: #GraceScanner).
	self removeGeneratedMethodsFrom: (Smalltalk globals at: #GraceScanner) class.
```
```
removeGeneratedMethodsFrom: aClass
	aClass organization protocols do: [  :p | 
		(p name includesSubstring: 'generated') ifTrue: [ 
			p methods do: [  :m |
			aClass removeSelector: m ]
		]
	]
```
The visitor shovel be removed too!   These cleanup methods are also handy to help prevent committing generated code to *git*.

I think that these methods should be generated by SmaCC, and (if they exist), executed by SmaCC before it generates new code.

**Comment by ThierryGoubier** [comment]
@apblack yes, that's a nice solution. I'll try to implement a variant of that.

How would that work if you have started to add your own methods to the AST nodes?

**Comment by apblack** [comment]
Manually-created methods in the AST nodes get removed by the above code, so it's not a good solution.  It seems to me that one has to get the hierarchy right before one starts adding code.  This is a problem, because I don't really understand what the hierarchy should be until after I've started adding code, realize that I have duplication, and want to eliminate it. 

Probably the right solution is a better implementation of %hierarchy, which will use the RB to move the existing subclasses around.

**Comment by apblack** [comment]
I've been working around this issue by first making changes to the %hierarchy directives, and then refactoring my existing node classes to match.  This works, but is error-prone.

If it's too much to expect _SmaCC_ to refactor my node hierarchy for me, perhaps a compromise is for it to _check_ that the existing hierarchy (if there is one) corresponds to the `%hierarchy` directives, and produce a warning message if it does not.

**Comment by apblack** [comment]
I've been using a little bit of code that I wrote with @ToshRaka to work around this problem.  It parses the grammar, extracts the hierarchy declarations, and compares them with the hierarchy implemented by the node classes in the image.  If the two hierarchies are not the same, it complains.

I run this as part of my test suite, which I run every time that I change the grammar.  It would be even better if SmaCC did the same checks — after all, it has already parsed the grammar.

[My code is here](https://github.com/apblack/GraceInPharo/tree/master/Grace-HValidator.package); you are welcome to incorporate it into SmaCC if you agree that this check would be useful.
